src/model/introspect.rs

```diff
@@ -84,6 +85,8 @@ impl StabilityProvider {
 pub struct Introspector {
     pages: usize,
     elems: Vec<(Content, Position)>,
+    /// A small cache of positions for faster lookups in nested queries.
+    position_cache: HashMap<Location, usize>,
```

Comment:
it's not really a cache if you intialize it at the beginning. also, its about locations not positions. maybe just name it `locations`?